File: cpp/tools/behavior_oracle.cpp
// Emits the C++ ECS server's observable derived gameplay values.  The Node
// parity runner compares these JSON tuples with behaviorOracle.ts.
#include <cstdlib>
#include <iostream>
#include <stdexcept>
#include <string>

#include "shared/core/json.h"
#include "shared/game/config.h"
#include "shared/game/rarity.h"
#include "server/systems/petals.h"
#include "server/systems/combat.h"

namespace {

using namespace flix;

void emit(const std::string& path, double value) {
    Json row = Json::array();
    row.push(path);
    row.push(value);
    std::cout << row.dump() << '\n';
}

void emit(const std::string& path, const std::string& value) {
    Json row = Json::array();
    row.push(path);
    row.push(value);
    std::cout << row.dump() << '\n';
}

void emit(const std::string& path, bool value) {
    Json row = Json::array();
    row.push(path);
    row.push(value);
    std::cout << row.dump() << '\n';
}

const char* aiName(AiKind ai) {
    switch (ai) {
        case AiKind::Passive: return "passive";
        case AiKind::Neutral: return "neutral";
        case AiKind::Hostile: return "hostile";
        case AiKind::Sandstorm: return "sandstorm";
        case AiKind::Stationary: return "stationary";
    }
    return "neutral";
}

struct EquippedPetal {
    const char* id;
    Rarity rarity;
    int slot;
};

void emitModifierScenario(const char* name, const ContentRegistry& registry,
                          std::initializer_list<EquippedPetal> equipped) {
    World world;
    CommandBuffer commands{world};
    const Entity player = world.create();
    world.add<PlayerTag>(player);
    world.add<Transform>(player, Transform{{1000.0, 1000.0}, 0.0});
    world.add<Body>(player, Body{kPlayerBaseRadius, 1.0});
    world.add<Health>(player, Health{110.0, 110.0, 0.0, 0.0});
    world.add<PlayerProgress>(player, PlayerProgress{});
    world.add<PlayerModifiers>(player);
    world.add<PlayerInput>(player);
    world.add<Loadout>(player);
    world.add<PetalRing>(player);

    Loadout& loadout = world.get<Loadout>(player);
    for (const EquippedPetal& item : equipped) {
        LoadoutSlot& slot = loadout.slots[static_cast<std::size_t>(item.slot)];
        slot.configIndex = registry.petalIndex(item.id);
        slot.rarity = item.rarity;
    }

    PetalSystem petals;
    petals.run(world, registry, 1000.0, 1.0, commands);
    commands.flush();

    const PlayerModifiers& modifiers = world.get<PlayerModifiers>(player);
    const double rotationSpeed = world.get<PetalRing>(player).spin / kPetalSpinRate;
    const std::string prefix = std::string("scenario/modifiers/") + name;
    emit(prefix + "/damage", modifiers.damageScale);
    emit(prefix + "/max-health", modifiers.maxHealthScale);
    emit(prefix + "/speed", modifiers.speedScale);
    emit(prefix + "/range", modifiers.rangeScale);
    emit(prefix + "/rotation-speed", rotationSpeed);
    emit(prefix + "/player-radius", modifiers.sizeScale);
    emit(prefix + "/luck", modifiers.luck);
    emit(prefix + "/magnetism", modifiers.magnetism);
    emit(prefix + "/aggro-radius", modifiers.aggroRadiusBonus);
    emit(prefix + "/petal-attraction-radius", modifiers.petalAttractionRadius);
    emit(prefix + "/poison-armor", modifiers.poisonArmor);
}

void emitMobHealthScenario() {
    World world;
    CombatSystem combat;

    const Entity player = world.create();
    world.add<PlayerTag>(player);
    world.add<Faction>(player, Faction{Team::Players, false});

    const Entity mob = world.create();
    world.add<MobTag>(mob);
    world.add<Health>(mob, Health{100.0, 100.0, 0.0, 0.0});
    world.add<Faction>(mob, Faction{Team::Hostiles, false});
    world.add<Bounty>(mob, Bounty{});

    combat.applyDamage(world, mob, player, 30.0, 1000.0);
    emit("scenario/mob-health/positive-hit", world.get<Health>(mob).current);
    combat.applyDamage(world, mob, player, -10.0, 1001.0);
    emit("scenario/mob-health/negative-hit", world.get<Health>(mob).current);
    combat.applyDamage(world, mob, player, 500.0, 1002.0);
    emit("scenario/mob-health/lethal-hit", world.get<Health>(mob).current);
    emit("scenario/mob-health/dead", world.has<Dead>(mob));
}

} // namespace

int main(int argc, char** argv) {
    using namespace flix;
    try {
        if (argc != 4) {
            std::cerr << "usage: behavior_oracle <mobs.json> <petals.json> <mob_xp.json>\n";
            return 2;
        }

        ContentRegistry registry;
        std::string error;
        if (!registry.loadFiles(argv[1], argv[2], argv[3], error)) {
            std::cerr << error << '\n';
            return 2;
        }

        for (int i = 0; i < kRarityCount; ++i) {
            const Rarity rarity = static_cast<Rarity>(i);
            const std::string prefix = std::string("rarity/") + rarityName(rarity);
            emit(prefix + "/craft-percent", craftSuccessChance(rarity) * 100.0);
            emit(prefix + "/drop-upgrade-percent", dropUpgradeChance(rarity) * 100.0);
            emit(prefix + "/drop-downgrade-fraction", dropDowngradeChance(rarity));
            for (int j = 0; j < kRarityCount; ++j) {
                const Rarity target = static_cast<Rarity>(j);
                emit(prefix + "/stall/" + rarityName(target), stallPower(rarity, target));
            }
        }

        for (std::size_t i = 0; i < registry.mobCount(); ++i) {
            const auto index = static_cast<std::uint16_t>(i);
            const MobConfig& config = registry.mob(index);
            for (int tier = 0; tier < kRarityCount; ++tier) {
                const Rarity rarity = static_cast<Rarity>(tier);
                const MobStats stats = registry.mobStats(index, rarity);
                const std::string prefix = "mob/" + config.id + "/" + rarityName(rarity);
                emit(prefix + "/health", stats.health);
                emit(prefix + "/damage", stats.damage);
                emit(prefix + "/radius", stats.radius);
                emit(prefix + "/mass", stats.mass);
                emit(prefix + "/xp", stats.xp);
                emit(prefix + "/aggro-range", stats.aggroRange);
                emit(prefix + "/attack-cooldown-ms", stats.attackCooldownMillis);
                emit(prefix + "/poison-per-second", stats.poisonPerSecond);
                emit(prefix + "/poison-duration-ms", stats.poisonDurationMillis);
                emit(prefix + "/visual-scale", stats.visualScale);
                emit(prefix + "/spawn-weight", stats.spawnWeight);
                emit(prefix + "/ai", std::string(aiName(stats.ai)));
                emit(prefix + "/ambient", stats.ambient ? 1.0 : 0.0);
            }
        }

        for (std::size_t i = 0; i < registry.petalCount(); ++i) {
            const auto index = static_cast<std::uint16_t>(i);
            const PetalConfig& config = registry.petal(index);
            for (int tier = 0; tier < kRarityCount; ++tier) {
                const Rarity rarity = static_cast<Rarity>(tier);
                const PetalStats stats = registry.petalStats(index, rarity);
                const PetalModifiers& modifiers = stats.modifiers;
                const std::string prefix = "petal/" + config.id + "/" + rarityName(rarity);
                emit(prefix + "/damage", stats.damage);
                emit(prefix + "/health", stats.health);
                emit(prefix + "/reload-ms", stats.reloadMillis);
                emit(prefix + "/poison-per-second", stats.poisonPerSecond);
                emit(prefix + "/poison-duration-ms", stats.poisonDurationMillis);
                emit(prefix + "/burst-heal", stats.heal);
                emit(prefix + "/burst-heal-charge-ms", stats.healChargeMillis);
                emit(prefix + "/passive-heal-per-second", stats.passiveHealPerSecond);
                emit(prefix + "/max-mana", stats.maxMana);
                emit(prefix + "/burst-mana", stats.mana);
                emit(prefix + "/burst-mana-charge-ms", stats.manaChargeMillis);
                emit(prefix + "/passive-mana-per-second", stats.passiveManaPerSecond);
                emit(prefix + "/required-mana", stats.requiredMana);
                emit(prefix + "/knockback", stats.knockback);
                emit(prefix + "/shield", stats.shield);
                emit(prefix + "/slow-factor", stats.slowFactor);
                emit(prefix + "/slow-duration-ms", stats.slowDurationMillis);
                emit(prefix + "/web-radius", stats.webRadius);
                emit(prefix + "/sponge-duration-ms", stats.spongeDamageDurationMillis);
                emit(prefix + "/attraction-force", stats.attractionForce);
                emit(prefix + "/size", stats.size);
                emit(prefix + "/count", static_cast<double>(stats.count));
                emit(prefix + "/camera-zoom", stats.cameraZoom);
                emit(prefix + "/modifier/damage", modifiers.damage);
                emit(prefix + "/modifier/max-health", modifiers.maxHealth);
                emit(prefix + "/modifier/speed", modifiers.speed);
                emit(prefix + "/modifier/range", modifiers.range);
                emit(prefix + "/modifier/rotation-speed", modifiers.rotationSpeed);
                emit(prefix + "/modifier/player-radius", modifiers.playerRadius);
                emit(prefix + "/modifier/luck", modifiers.luck);
                emit(prefix + "/modifier/magnetism", modifiers.magnetism);
                emit(prefix + "/modifier/aggro-radius", modifiers.aggroRadius);
                emit(prefix + "/modifier/petal-attraction-radius", modifiers.petalAttractionRadius);
                emit(prefix + "/modifier/poison-armor", modifiers.poisonArmor);
            }
        }

        emitModifierScenario("empty", registry, {});
        emitModifierScenario("mixed-active", registry, {
            {"faster", Rarity::Rare, 0}, {"powder", Rarity::Epic, 1},
            {"soil", Rarity::Legendary, 2}, {"air", Rarity::Uncommon, 3},
            {"clover", Rarity::Mythic, 4}, {"lotus", Rarity::Ultra, 5},
            {"lentil", Rarity::Super, 6}, {"basic", Rarity::Common, 7},
            {"faster", Rarity::Common, 9},
        });
        emitModifierScenario("storage-ignored", registry, {
            {"basic", Rarity::Common, 0}, {"faster", Rarity::Apex, 10},
            {"soil", Rarity::Apex, 11}, {"clover", Rarity::Apex, 12},
            {"lentil", Rarity::Apex, 13},
        });
        emitMobHealthScenario();
        // There used to be a movement scenario here, walked over the tile grid
        // in src/map_bundle.ts so both builds could step the same walls. The
        // map is a Tiled file now and the frozen TypeScript build cannot read
        // one, so there is no map both sides can agree on any more -- and an
        // oracle that stepped a DIFFERENT map would report a parity failure
        // that means nothing. Movement against real terrain is covered by
        // cpp/tests/movement_tests.cpp instead; behavior-parity.js skips the
        // keys the TypeScript side still emits for it, and says why.
    } catch (const std::exception& ex) {
        std::cerr << ex.what() << '\n';
        return 2;
    }
    return 0;
}

File: src/server/behaviorOracle.ts
/*
 * Machine-readable behavioral oracle for the native server.
 *
 * Keep this module side-effect free: the parity runner executes it in a fresh
 * process and compares every emitted observation with the C++ ECS server's
 * probe.  Each line is a JSON tuple so content ids can safely contain any
 * punctuation without inventing an escaping convention.
 */
import '../map_data'; // Populate constants.WALL_GRID exactly as the live TS server does.
import { getAllMobTypes, getMobStats, RARITY_LEVELS as MOB_RARITIES } from '../mobs';
import {
    getAllPetalTypes,
    getEffectivePetalCooldown,
    getPetalStats,
    RARITY_LEVELS as PETAL_RARITIES,
} from '../petals';
import {
    RARITY_ORDER,
    getCraftingChance,
    getDropDowngradeChance,
    getDropUpgradeChance,
    stallPower,
} from './shared/rarity';
import { calculatePlayerModifiers } from './shared/playerModifiers';
import { ServerPlayer } from '../player';
import { World } from '../ecs/world';
import * as C from '../ecs/components';
import { damageMob } from './mobFields';
import {
    PlayerMoveState, stepPlayerMovement, WALL_GRID, getTileState,
    resolveEntityWallCollisions,
} from '../constants';

type Metric = number | string | boolean;

function emit(path: string, value: Metric): void {
    process.stdout.write(`${JSON.stringify([path, value])}\n`);
}

function numberOr(value: number | undefined | null, fallback: number): number {
    return value === undefined || value === null ? fallback : value;
}

function sectionMask(sections: number[]): number {
    let mask = 0;
    for (const section of sections) {
        if (section >= 0 && section < 9) mask |= 1 << section;
    }
    return mask;
}

for (let i = 0; i < RARITY_ORDER.length; i++) {
    const rarity = RARITY_ORDER[i];
    emit(`rarity/${rarity}/craft-percent`, getCraftingChance(i));
    emit(`rarity/${rarity}/drop-upgrade-percent`, getDropUpgradeChance(rarity));
    emit(`rarity/${rarity}/drop-downgrade-fraction`, getDropDowngradeChance(rarity));
    for (const target of RARITY_ORDER) {
        emit(`rarity/${rarity}/stall/${target}`, stallPower(rarity, target));
    }
}

for (const mobType of getAllMobTypes()) {
    for (const rarity of MOB_RARITIES) {
        const stats = getMobStats(mobType, rarity);
        if (!stats) throw new Error(`missing TypeScript mob stats: ${mobType}/${rarity}`);
        const prefix = `mob/${mobType}/${rarity}`;
        emit(`${prefix}/health`, stats.health);
        emit(`${prefix}/damage`, stats.damage);
        emit(`${prefix}/radius`, stats.size * 20);
        emit(`${prefix}/mass`, stats.mass);
        emit(`${prefix}/xp`, stats.xp);
        emit(`${prefix}/aggro-range`, stats.range);
        emit(`${prefix}/attack-cooldown-ms`, stats.cooldown);
        emit(`${prefix}/poison-per-second`, numberOr(stats.poison, 0) * 1000);
        emit(`${prefix}/poison-duration-ms`, numberOr(stats.poisonDuration, 0));
        emit(`${prefix}/visual-scale`, numberOr(stats.visual_scale, 1));
        emit(`${prefix}/spawn-weight`, stats.spawn_weight);
        emit(`${prefix}/ai`, stats.ai_type);
        emit(`${prefix}/section-mask`, sectionMask(stats.section));
    }
}

for (const petalType of getAllPetalTypes()) {
    for (const rarity of PETAL_RARITIES) {
        const stats = getPetalStats(petalType, rarity);
        if (!stats) throw new Error(`missing TypeScript petal stats: ${petalType}/${rarity}`);
        const prefix = `petal/${petalType}/${rarity}`;
        const modifiers = stats.playerModifiers;
        emit(`${prefix}/damage`, stats.damage);
        emit(`${prefix}/health`, stats.health);
        emit(`${prefix}/reload-ms`, getEffectivePetalCooldown(petalType, rarity, stats));
        emit(`${prefix}/poison-per-second`, numberOr(stats.poison, 0) * 1000);
        emit(`${prefix}/poison-duration-ms`, numberOr(stats.poisonDuration, 0));
        emit(`${prefix}/burst-heal`, numberOr(stats.burstHeal, 0));
        emit(`${prefix}/burst-heal-charge-ms`, numberOr(stats.burstHealChargeMs, 0));
        emit(`${prefix}/passive-heal-per-second`, numberOr(stats.passiveHeal, 0));
        emit(`${prefix}/knockback`, numberOr(stats.knockback, 5));
        emit(`${prefix}/shield`, numberOr(stats.burstShield, 0));
        emit(`${prefix}/slow-factor`, numberOr(stats.slowFactor, 1));
        emit(`${prefix}/slow-duration-ms`, numberOr(stats.slowDuration, 0));
        emit(`${prefix}/web-radius`, numberOr(stats.webRadius, 0));
        emit(`${prefix}/sponge-duration-ms`, numberOr(stats.spongeDamageDuration, 0));
        emit(`${prefix}/attraction-force`, numberOr(stats.attractionForce, 0));
        emit(`${prefix}/size`, stats.size);
        emit(`${prefix}/count`, numberOr(stats.count, 1));
        emit(`${prefix}/camera-zoom`, numberOr(stats.cameraZoom, 1));
        emit(`${prefix}/modifier/damage`, numberOr(modifiers?.damage, 1));
        emit(`${prefix}/modifier/max-health`, numberOr(modifiers?.maxHealth, 1));
        emit(`${prefix}/modifier/speed`, numberOr(modifiers?.speed, 1));
        emit(`${prefix}/modifier/range`, numberOr(modifiers?.range, 1));
        emit(`${prefix}/modifier/rotation-speed`, numberOr(modifiers?.rotationSpeed, 1));
        emit(`${prefix}/modifier/player-radius`, numberOr(modifiers?.playerRadius, 1));
        emit(`${prefix}/modifier/luck`, numberOr(modifiers?.luck, 0));
        emit(`${prefix}/modifier/magnetism`, numberOr(modifiers?.magnetism, 0));
        emit(`${prefix}/modifier/aggro-radius`, numberOr(modifiers?.aggroRadius, 0));
        emit(`${prefix}/modifier/petal-attraction-radius`, numberOr(modifiers?.petalAttractionRadius, 0));
        emit(`${prefix}/modifier/poison-armor`, numberOr(modifiers?.poisonArmor, 0));
    }
}

const modifierScenarios: Array<{ name: string; loadout: Array<[string, string] | null> }> = [
    { name: 'empty', loadout: [] },
    {
        name: 'mixed-active',
        loadout: [
            ['faster', 'rare'], ['powder', 'epic'], ['soil', 'legendary'],
            ['air', 'uncommon'], ['clover', 'mythic'], ['lotus', 'ultra'],
            ['lentil', 'super'], ['basic', 'common'], null, ['faster', 'common'],
        ],
    },
    {
        name: 'storage-ignored',
        loadout: [
            ['basic', 'common'], null, null, null, null, null, null, null, null, null,
            ['faster', 'apex'], ['soil', 'apex'], ['clover', 'apex'], ['lentil', 'apex'],
        ],
    },
];

for (const scenario of modifierScenarios) {
    const loadout = scenario.loadout.map(item => item === null ? null : ({
        type: 'petal', petalType: item[0], rarity: item[1],
    }));
    const modifiers = calculatePlayerModifiers({ loadout } as unknown as ServerPlayer);
    const prefix = `scenario/modifiers/${scenario.name}`;
    emit(`${prefix}/damage`, numberOr(modifiers.damage, 1));
    emit(`${prefix}/max-health`, numberOr(modifiers.maxHealth, 1));
    emit(`${prefix}/speed`, numberOr(modifiers.speed, 1));
    emit(`${prefix}/range`, numberOr(modifiers.range, 1));
    emit(`${prefix}/rotation-speed`, numberOr(modifiers.rotationSpeed, 1));
    emit(`${prefix}/player-radius`, numberOr(modifiers.playerRadius, 1));
    emit(`${prefix}/luck`, numberOr(modifiers.luck, 1));
    emit(`${prefix}/magnetism`, numberOr(modifiers.magnetism, 0));
    emit(`${prefix}/aggro-radius`, numberOr(modifiers.aggroRadius, 0));
    emit(`${prefix}/petal-attraction-radius`, numberOr(modifiers.petalAttractionRadius, 30));
    emit(`${prefix}/poison-armor`, numberOr(modifiers.poisonArmor, 0));
}

// Drive the real ECS health writer, including glitch's intentionally negative
// damage.  This catches semantic differences that matching stat tables alone
// cannot: TypeScript negative mob damage heals and may exceed the old health.
{
    const world = new World();
    const mob = world.create();
    world.add(mob, C.Health, { current: 100, max: 100 });
    emit('scenario/mob-health/positive-hit', damageMob(mob, 30, world));
    emit('scenario/mob-health/negative-hit', damageMob(mob, -10, world));
    emit('scenario/mob-health/lethal-hit', damageMob(mob, 500, world));
    emit('scenario/mob-health/dead', world.has(mob, C.IsDead));
}

// Run the authoritative TypeScript movement step at the production 30 Hz.
// The chosen garden patch is open in the shared map, so this isolates input,
// friction, integration, body-size handling, and coast-down.
{
    const samples: Array<[string, number, number]> = [
        ['center', 3000, 3000], ['above', 3000, 2990], ['below', 3000, 3010],
        ['right', 3070, 3000], ['upper-right', 3070, 2990],
    ];
    for (const [name, x, y] of samples) {
        emit(`scenario/movement/tile/${name}`, getTileState(WALL_GRID, x, y));
    }
    const resolvedStart = resolveEntityWallCollisions(3000, 2940, 20);
    emit('scenario/movement/resolve-start/x', resolvedStart.x);
    emit('scenario/movement/resolve-start/y', resolvedStart.y);
    let state: PlayerMoveState = { x: 3000, y: 2940, vx: 0, vy: 0 };
    const dt = 1 / 30;
    const phases: Array<[string, number, number]> = [
        ['wall-slide', 300 / Math.sqrt(2), 300 / Math.sqrt(2)],
        ['half-up', 0, -150],
        ['release', 0, 0],
    ];
    for (const [name, targetVX, targetVY] of phases) {
        for (let tick = 0; tick < 10; tick++) {
            state = stepPlayerMovement(state, targetVX, targetVY, dt, 40);
            emit(`scenario/movement/${name}/tick-${tick + 1}/x`, state.x);
            emit(`scenario/movement/${name}/tick-${tick + 1}/y`, state.y);
            emit(`scenario/movement/${name}/tick-${tick + 1}/vx`, state.vx);
            emit(`scenario/movement/${name}/tick-${tick + 1}/vy`, state.vy);
        }
        emit(`scenario/movement/${name}/x`, state.x);
        emit(`scenario/movement/${name}/y`, state.y);
        emit(`scenario/movement/${name}/vx`, state.vx);
        emit(`scenario/movement/${name}/vy`, state.vy);
    }
}
